src/databricks/labs/ucx/source_code/notebooks/sources.py

```diff
@@ -93,35 +112,156 @@ class NotebookLinter:
+        for cell in notebook.cells:
+            if isinstance(cell, RunCell):
+                yield from self._load_tree_from_run_cell(cell)
+                continue
+            if isinstance(cell, PythonCell):
+                yield from self._load_tree_from_python_cell(cell, register_trees)
+                continue
+
+    def _load_tree_from_python_cell(self, python_cell: PythonCell, register_trees: bool) -> Iterable[Advice]:
+        try:
+            tree = Tree.normalize_and_parse(python_cell.original_code)
+            if register_trees:
+                self._python_trees[python_cell] = tree
+            yield from self._load_children_from_tree(tree)
+        except AstroidSyntaxError as e:
+            yield Failure('syntax-error', str(e), 0, 0, 0, 0)
+
+    def _load_children_from_tree(self, tree: Tree) -> Iterable[Advice]:
+        assert isinstance(tree.node, Module)
+        # look for child notebooks (and sys.path changes that might affect their loading)
+        base_nodes: list[NodeBase] = []
+        base_nodes.extend(DbutilsLinter.list_dbutils_notebook_run_calls(tree))
+        base_nodes.extend(SysPathChange.extract_from_tree(self._session_state, tree))
+        _ = """ run_magics = self._list_run_magics(tree)
+        # magic_commands, _ = MagicCommand.extract_from_tree(tree, self._report_problem)
+        base_nodes.extend(magic_commands)
+        """
+        if len(base_nodes) == 0:
+            self._python_linter.append_tree(tree)
+            return
+        # append globals
+        globs = cast(Module, tree.node).globals
+        self._python_linter.append_globals(globs)
+        # need to execute things in intertwined sequence so concat and sort them
+        nodes = list(cast(Module, tree.node).body)
+        base_nodes = sorted(base_nodes, key=lambda node: (node.node.lineno, node.node.col_offset))
+        yield from self._load_children_with_base_nodes(nodes, base_nodes)
+        # append remaining nodes
+        self._python_linter.append_nodes(nodes)
+
+    def _load_children_with_base_nodes(self, nodes: list[NodeNG], base_nodes: list[NodeBase]):
+        for base_node in base_nodes:
+            yield from self._load_children_with_base_node(nodes, base_node)
+
+    def _load_children_with_base_node(self, nodes: list[NodeNG], base_node: NodeBase):
+        while len(nodes) > 0:
+            node = nodes.pop(0)
+            self._python_linter.append_nodes([node])
+            if node.lineno < base_node.node.lineno:
+                continue
+            yield from self._load_children_from_base_node(base_node)
+
+    def _load_children_from_base_node(self, base_node: NodeBase):
+        if isinstance(base_node, SysPathChange):
+            yield from self._mutate_path_lookup(base_node)
+            return
+        if isinstance(base_node, NotebookRunCall):
+            has_unresolved, paths = base_node.get_notebook_paths(self._session_state)
```

Comment:
are you 100% sure this behavior is necessary? dbutils.notebook.run() does not propagate variables to the parent notebook, only `%run` does. provide a screenshot from a real workspace environment to confirm this.

Reply:
you're right, that was an incorrect assumption